qiskit/qasm3/exporter.py

```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
+        "ry": library.RYGate(*_FIXED_PARAMETERS[:1]),
+        "rz": library.RZGate(*_FIXED_PARAMETERS[:1]),
+        "cx": library.CXGate(),
+        "cy": library.CYGate(),
+        "cz": library.CZGate(),
+        "cp": library.CPhaseGate(*_FIXED_PARAMETERS[:1]),
+        "crx": library.CRXGate(*_FIXED_PARAMETERS[:1]),
+        "cry": library.CRYGate(*_FIXED_PARAMETERS[:1]),
+        "crz": library.CRZGate(*_FIXED_PARAMETERS[:1]),
+        "ch": library.CHGate(),
+        "swap": library.SwapGate(),
+        "ccx": library.CCXGate(),
+        "cswap": library.CSwapGate(),
+        "cu": library.CUGate(*_FIXED_PARAMETERS[:4]),
+        "CX": library.CXGate(),
+        "phase": library.PhaseGate(*_FIXED_PARAMETERS[:1]),
+        "cphase": library.CPhaseGate(*_FIXED_PARAMETERS[:1]),
+        "id": library.IGate(),
+        "u1": library.U1Gate(*_FIXED_PARAMETERS[:1]),
+        "u2": library.U2Gate(*_FIXED_PARAMETERS[:2]),
+        "u3": library.U3Gate(*_FIXED_PARAMETERS[:3]),
+    },
+}
+
+_BUILTIN_GATES = {
+    "U": library.UGate(*_FIXED_PARAMETERS[:3]),
+}
+
+
+@dataclasses.dataclass
+class GateDefinition:
+    """Symbol-table information on a gate."""
+
+    source: Gate | None
+    """The definition source of the gate.  This can be ``None`` if the gate was an overridden "basis
+    gate" for this export."""
+    node: ast.QuantumGateDefinition | None
+    """An AST node containing the gate definition.  This can be ``None`` if the gate came from an
+    included file, or is an overridden "basis gate" of the export."""
+
+
+class SymbolTable:
+    """Track Qiskit objects and the OQ3 identifiers used to refer to them."""
+
+    def __init__(self):
+        self.gates: collections.OrderedDict[str, GateDefinition | None] = {}
+        """Mapping of the symbol name to the "definition source" of the gate, which provides its
+        signature and decomposition.  The definition source can be `None` if the user set the gate
+        as a custom "basis gate".
+
+        Gates can only be declared in the global scope, so there is just a single look-up for this.
+
+        This is insertion ordered, and that can be relied on for iteration later."""
+        self.standard_gate_idents: dict[StandardGate, ast.Identifier] = {}
+        """Mapping of standard gate enumeration values to the identifier we represent that as."""
+        self.user_gate_idents: dict[int, ast.Identifier] = {}
+        """Mapping of `id`s of user gates to the identifier we use for it."""
+
+        self.variables: list[dict[str, object]] = [{}]
+        """Stack of mappings of variable names to the Qiskit object that represents them.
+
+        The zeroth index corresponds to the global scope, the highest index to the current scope."""
+        self.objects: list[dict[object, ast.Identifier]] = [{}]
+        """Stack of mappings of Qiskit objects to the identifier (or subscripted identifier) that
+        refers to them.  This is similar to the inverse mapping of ``variables``.
```

Comment:
(See also the previous comment on descriptions of mappings)
I'm looking for a way to describe a map without too much ambiguity and subject/verb disagreement. Maybe something like:
```suggestion
        """Mapping from standard gate enumeration values to the identifiers we represent them as."""
        self.user_gate_idents: dict[int, ast.Identifier] = {}
        """Mapping from `id`s of user gates to the identifiers we use for them."""

        self.variables: list[dict[str, object]] = [{}]
        """Stack of mappings from variable names to the Qiskit objects that represent them.

        The zeroth index corresponds to the global scope, the highest index to the current scope."""
        self.objects: list[dict[object, ast.Identifier]] = [{}]
        """Stack of mappings from Qiskit objects to the identifiers (or subscripted identifiers) that
        refer to them.  This is similar to the inverse mapping of ``variables``.
```

Reply:
Leaving as is, per https://github.com/Qiskit/qiskit/pull/12776#issuecomment-2242631327.